src/Raven.Server/Documents/Sharding/Background/ShardedDocumentsMigrator.cs

```diff
@@ -0,0 +1,106 @@
+        private readonly CancellationToken _token;
+        private readonly Logger _logger;
+
+        public ShardedDocumentsMigrator(ShardedDocumentDatabase database)
+        {
+            _database = database;
+            _token = database.ShardedDocumentsStorage.DocumentDatabase.DatabaseShutdown;
+            _logger = LoggingSource.Instance.GetLogger(database.Name, GetType().FullName);
+        }
+
+        internal async Task ExecuteMoveDocumentsAsync()
+        {
+            try
+            {
+                int bucket = -1;
+                int moveToShard = -1;
+                bool found = false;
+                bool monitorTaken = false;
+                try
+                {
+                    _token.ThrowIfCancellationRequested();
+
+                    Monitor.TryEnter(this, 0, ref monitorTaken);
+                    if (monitorTaken == false)
+                        return;
+
+                    if (_database.ServerStore.Sharding.HasActiveMigrations(_database.ShardedDatabaseName))
+                        return;
+
+                    using (_database.ShardedDocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+                    using (context.OpenReadTransaction())
+                    {
+                        var configuration = _database.ShardingConfiguration;
+                        for (var index = 0; index < configuration.BucketRanges.Count; index++)
+                        {
+                            var range = configuration.BucketRanges[index];
+                            if (range.ShardNumber == _database.ShardNumber)
+                                continue;
+
+                            var start = range.BucketRangeStart;
+                            var end = index == configuration.BucketRanges.Count - 1
+                                ? int.MaxValue
+                                : configuration.BucketRanges[index + 1].BucketRangeStart - 1;
+
+                            var bucketStatistics = ShardedDocumentsStorage.GetBucketStatistics(context, start, end);
+
+                            foreach (var bucketStats in bucketStatistics)
+                            {
+                                if (bucketStats.NumberOfDocuments == 0)
+                                    continue;
+
+                                bucket = bucketStats.Bucket;
+                                moveToShard = range.ShardNumber;
+                                found = true;
+                                break;
+                            }
+
+                            if (found)
+                                break;
+                        }
+                    }
+                }
+                finally
+                {
+                    if (monitorTaken)
+                        Monitor.Exit(this);
+                }
+
+                if (found)
+                    await MoveDocumentsToShardAsync(bucket, moveToShard);
```

Comment:
```suggestion
                    return MoveDocumentsToShardAsync(bucket, moveToShard);
```

Reply:
Not sure about that, there is a try catch block here?